Tutorial for changefeeds to confluent cloud kafka & SR

kathancox opened the pull request:
Fixes [DOC-3671](https://cockroachlabs.atlassian.net/browse/DOC-3671)

Added a tutorial to the CDC documentation on connecting a changefeed to Confluent Cloud Kafka with a Schema Registry

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/14732)
<!-- Reviewable:end -->

amruss commented on v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md:
```diff
@@ -0,0 +1,267 @@
+
+You will need the following set up before starting this tutorial:
+
+- A CockroachDB cluster. You can use a {{ site.data.products.db }} or {{ site.data.products.core }} cluster. For {{ site.data.products.db }}, see the [Quickstart with CockroachDB](../cockroachcloud/quickstart.html) guide. Or, see the {{ site.data.products.core }} [install](install-cockroachdb-mac.html) page. 
```
i think maybe we should add serverless? Since that is an option now too

amruss commented on v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md:
```diff
@@ -0,0 +1,267 @@
+SET CLUSTER SETTING kv.rangefeed.enabled = true;
+~~~
+
+Before running the [`CREATE CHANGEFEED`](create-changefeed.html) statement, you must **URL-encode** both the cluster's and the Schema Registry's API secret key.
```
Maybe we should link to a place that will encode for you? I always use this website: https://www.base64encoder.io/

amruss reviewed (approved):
Had a couple of suggestions, but overall this is great! Very thorough and easy to follow - love it

andf-crl reviewed (commented):
Nice one -- I got stuck on step 6, and have added my troubleshooting steps. I can take another look tomorrow -- the error could well be on my end -- but here are my comments as they stand today. Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/14732#-:-N92IBza6Xm5YC5b1e4b:bpkp0cc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @amruss, @andf-crl, and @kathancox)

___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 8 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N91zHeG6Nv4nHfrI-Oi:-N91zHeG6Nv4nHfrI-Oj:b-qu082m) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L8)):*
> ```Markdown
> ---
> 
> CockroachDB {{ site.data.products.enterprise }} changefeeds can stream change data out to [Apache Kafka](changefeed-sinks.html) with different [configuration settings](changefeed-sinks.html#kafka-sink-configuration) and [options](create-changefeed.html). [Confluent Cloud](https://www.confluent.io/confluent-cloud/) provides a fully managed service for running Apache Kafka as well as the [Confluent Cloud Schema Registry](https://docs.confluent.io/platform/current/schema-registry/index.html). 
> ```

nit: `[Apache Kafka](changefeed-sinks.html)` could be `[Apache Kafka](changefeed-sinks.html#kafka)`. Alternatively, we could link out to https://kafka.apache.org/ as with the confluent links: this intro paragraph seems like the right place to establish the external components of this tutorial.
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 40 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N923GjzCnSKQ58KWdt2:-N923GjzCnSKQ58KWdt3:b-v5zyzo) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L40)):*
Suggest: something like:

> These instructions use the `--save` flag to store your username and password to a local file for convenience during this tutorial, but you can omit this flag if you would prefer to manually authenticate yourself each time.
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 56 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N924efOB63onpkS4VeT:-N924efOB63onpkS4VeU:bq8c44j) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L56)):*
Suggest: a clarification between `ID` and `Name`, as `use` cannot take the name. Perhaps:

> Set the environment that you would like to create your cluster in, using the environment's `ID`, which starts with `env-`:
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 70 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N925oVK9Rh-qY4pmZRD:-N925oVK9Rh-qY4pmZRE:bqsr0jc) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L70)):*
> ```Markdown
> Here the name of the cluster is `movr-confluent-tutorial`, but you can change this for your cluster.
> 
> Note that the `--cloud` and `--region` flags are **required** when running this `create` command. See Confluent's documentation on [`confluent kafka cluster create`](https://docs.confluent.io/confluent-cli/current/command-reference/kafka/cluster/confluent_kafka_cluster_create.html).
> ```

nit: `when running this` ->. `when running the`
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 72 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N9265LvC-Y7UwKF2MG6:-N9265Lw3gzFpPkkz4H5:b-2kmw8j) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L72)):*
Suggestion: something like:

> The `create` command returns your new cluster's details, with a format similar to the following:

Then see my next comment for "need some info later" content.

nit: this might be a good spot to advise that Kafka cluster creation can take up to 5 minutes, per the returned message. You might highlight status `PROVISIONING` as part of this. Certainly optional.
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 92 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N926oT71bpQg7oSVrDI:-N926oT71bpQg7oSVrDJ:bxkxcgc) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L92)):*
Suggestion: What do you think of moving "we'll need this data later" here, something like:

> You'll need this information later in the tutorial, but you can also access this status at any time with the following command:
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 137 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N928aKo34t7GABETKJe:-N928aKo34t7GABETKJf:b-gorptp) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L137)):*
nit: if we use `{placeholder}` notation, we should use highly-generic naming as well, perhaps:

> `{topic}`

`{users}` seems like an actual, rational target topic name, but as-is the command 500 errors (curly braces). A more generic name here helps to convey: supply your own name, remove the curly braces.

EDIT: ah we are using `users` throughout the tutorial as the specific target topic. In that case, I'd recommend removing the `{}` from it here.
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 140 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N929FoI1448iUQG14_s:-N929FoI1448iUQG14_t:b-pc7305) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L140)):*
> ```Markdown
> ~~~
> 
> `users` will be the topic name for this tutorial. Change the topic name for your purposes and run the previous command for each topic you would like to create.
> ```

nit: `Change the topic name ...` (i.e. a mandatory directive) --> `If needed, you can change the topic name ...` (an optional consideration)
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 143 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N929SE18w8ygL1uEItP:-N929SE18w8ygL1uEItQ:brnxsj3) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L143)):*
> ```Markdown
> 
> {{site.data.alerts.callout_success}}
> If you are using a Dedicated Confluent Cloud cluster, you can auto enable topic creation. For further detail, see [Enable automatic topic creation](https://docs.confluent.io/cloud/current/clusters/broker-config.html#enable-automatic-topic-creation). 
> ```

This is a great tip!!

nit: `auto enable topic creation` --> `enable automatic topic creation`
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 169 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N929xjH4Sylfa5C-1JZ:-N929xjH4Sylfa5C-1J_:b-vrsqv4) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L169)):*
optional:

> from your output, which starts with `lsrc-`:
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 187 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N92ADzoBwnogAkrZKU3:-N92ADzoBwnogAkrZKU4:bx2f83y) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L187)):*
similar placeholder nit here: let's super-genericize placeholder values to prevent users from running as-is. Suggest:

> `--sr-endpoint {schema_registry_endpoint_url} \`
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 188 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N92EXVv4JPnASCvwHwd:-N92EXVv4JPnASCvwHwe:b60fe9m) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L188)):*
> ```Markdown
>  --from-beginning \
>  --sr-endpoint https://{psrc-x77pq.us-central1.gcp}.confluent.cloud \
>  --sr-api-key {SCHEMA REGISTRY API Key} \
> ```

nit: either `KEY` to match all caps elsewhere, or everywhere else to match exactly as returned from output (example: `{Secret}`). Easier to just go with: `KEY`
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 194 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N92Bv3-CTtsX22yGJpS:-N92Bv3-CTtsX22yGJpT:bgm7n7l) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L194)):*
> ```Markdown
> In this command, you need to pass the following Schema Registry details:
> 
> - The endpoint from the output in [Step 4](#step-4-create-a-confluent-schema-registry)
> ```

clarification: `endpoint URL`
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 197 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N92CJ9v-_sOb8BpTiM-:-N92CJ9v-_sOb8BpTiM0:b6ry02i) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L197)):*
Troubleshooting: I ran into issues here:

> Error: `Error: no API key selected for resource "{my cluster ID}"`

The error message suggested running: `confluent api-key use <API_KEY> --resource {my schema cluster ID}`, which resulted in:

> Error: `Error: unable to set active API key: invalid API key "{my schema API key}" for resource "{my schema cluster ID}"`

Suggestion: somewhere around here it might be useful to also add a little troubleshooting guidance

> Make sure `confluent kafka cluster describe {cluster_ID}` returns a `Status` of `UP`.
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 200 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N92C48L3axv66P2wL4n:-N92C48M3ZgIydmvD7ON:beblqgb) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L200)):*
> ```Markdown
> 
> {{site.data.alerts.callout_success}}
> Run `confluent schema-registry cluster describe` to access details for the Schema Registry.
> ```

clarity: `for the Schema Registry, if needed.`
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 203 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N92FCUS8O2B_js3wHEq:-N92FCUS8O2B_js3wHEr:b-52xi56) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L203)):*
> ```Markdown
> {{site.data.alerts.end}}
> 
> ## Step 7. Create a changefeed 
> ```

Note: I wasn't able to get to this step programmatically, so all remaining comments are just language.

nit: step 7 is rather large and spans several different concepts:
- Connect
- Enable rangefeeds
- base64 encode
- create changefeeds

I think splitting this into one or two additional steps would make sense.
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 210 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N92Fe2t-Ne0eBBgXDIV:-N92Fe2t-Ne0eBBgXDIW:b-jaf8o7) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L210)):*
> ```Markdown
> 
> ~~~shell
> cockroach sql --url {CONNECTION STRING}
> ```

Recommend a quick link to the connection string reference.
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 219 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N8xnmoLo26yKbxC8HJw-r1-219:-N92FmPE3RyAfasEjG7V:b-g640to) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L219)):*
<details><summary><i>Previously, amruss (Abigail Russell) wrote…</i></summary><blockquote>

Maybe we should link to a place that will encode for you? I always use this website: https://www.base64encoder.io/
</blockquote></details>

base64 and URL-encode are not quite the same. A bit of googling suggests that Kafka likely requires base64 encoding specifically, which should be what we specify here as well.
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 227 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N92HT1f888WmoDr9s4A:-N92HT1f888WmoDr9s4B:bzgpr2k) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L227)):*
> ```Markdown
> ~~~
> 
> To connect to the Kafka cluster, use the `Endpoint` from your cluster details preceded with the `kafka://` scheme. For example, an endpoint of `pkc-4yyd6.us-east1.gcp.confluent.cloud:9092` would be: `kafka://pkc-4yyd6.us-east1.gcp.confluent.cloud:9092`. 
> ```

nit: `Endpoint URL` to match outputted naming
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 240 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N92Hb_VAQ20d06vhJp6:-N92Hb_VAQ20d06vhJp7:brspyk9) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L240)):*
> ```Markdown
> 
> - `format = avro`
> - `confluent_schema_registry = "https://{API KEY}:{URL-ENCODED SCHEMA REGISTRY SECRET KEY}@{SCHEMA REGISTRY URL}:443"`. Note that the schema registry uses basic authentication, which means that the URL's format is different to the Kafka URL.
> ```

nit: `different to` --> `different from`
___
*[`v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md` line 245 at r1](https://reviewable.io/reviews/cockroachdb/docs/14732#-N92HmH43ttBpTjvYbe-:-N92HmH43ttBpTjvYbe0:bdtq1he) ([raw file](https://github.com/cockroachdb/docs/blob/ba4520f95d40fe4824a4240769caef89a030af88/v22.1/stream-a-changefeed-to-a-confluent-cloud-kafka-cluster.md#L245)):*
> ```Markdown
> ## Step 8. Verify the output
> 
> Move to the terminal window in which you started the consumer. As the changefeed runs, you will see the change data messages:
> ```

nit: `started the Kafka consumer.`

optional nit: `see change data messages similar to the following:` (because we recommend `movr` we could insist on these specific messages)


<!-- Sent from Reviewable.io -->

kathancox commented:
Hey @andf-crl , I will take a look through the feedback! I'm pretty sure this is URL-encoded — this is what I did using a URL-encoder and not base64. The format of the original secret keys with "/" "+" etc. make me think that URL-encoding to go to "%2F", for example, is the right encoding here. On your run through, did you base64 encode?

kathancox commented:
Hey @andf-crl, I have incorporated most of your comments. I have also added some clarifications and additions in some places that I felt were missing. I tested this again to see if I ran into your error. I didn't, but felt that some things could be clarified. The only things that I did not change, are the encoding per the previous comment and the ref to `Endpoint` and adding `URL` to it for the cluster. Annoyingly the output details for the SR and cluster refer to the endpoint differently. One with `Endpoint URL`, one just `Endpoint`.

Also, I had to force push after a rebase, so apologies if that messes up the diff view!

andf-crl reviewed (approved):
Thanks Kathryn! Good catch on `Endpoint | URL` output, you're absolutely right, I withdraw that one. I'll also defer to you and Abigail on base64.

I think this was likely my stumbling block:

> confluent api-key use {API KEY} --resource {CLUSTER ID}

Thanks for clarifying!

Copy changes look great, and this LGTM! Thanks for the helpful comments!

kathancox commented:
Thanks @andf-crl & @amruss for the reviews!
